core/backend/src/StashManager.ts

```diff
@@ -0,0 +1,458 @@
+import { ChangesetIdWithIndex, IModelError, LocalDirName, LockState } from "@itwin/core-common";
+import { existsSync, mkdirSync, readdirSync, statSync, unlinkSync } from "node:fs";
+import * as path from "path";
+import { BriefcaseManager } from "./BriefcaseManager";
+import { BriefcaseDb } from "./IModelDb";
+import { _elementWasCreated, _getHubAccess, _hubAccess, _nativeDb } from "./internal/Symbols";
+import { SQLiteDb } from "./SQLiteDb";
+import { TxnProps } from "./TxnManager";
+import { IModelHost } from "./IModelHost";
+import { BackendLoggerCategory } from "./BackendLoggerCategory";
+
+const loggerCategory = BackendLoggerCategory.StashManager;
+
+/**
+ * Custom error class for stash-related errors.
+ * @internal
+ */
+export class StashError extends Error {
+  constructor( message: string) {
+    super(message);
+    this.name = "StashError";
+    Logger.logError(loggerCategory, message);
+  }
+}
+
+/**
+ * Properties of a stash
+ * @internal
+ */
+export interface StashProps {
+  /** Unique identifier for the stash */
+  readonly id: GuidString;
+
+  /** ID of the iModel being stashed */
+  readonly iModelId: GuidString;
+
+  /** ID of the briefcase being stashed */
+  readonly briefcaseId: number;
+
+  /** ISO local Timestamp of the stash */
+  readonly timestamp: string;
+
+  /** Description of the stash */
+  readonly description: string;
+
+  /** Hash of the stash */
+  readonly hash: string;
+
+  /** Parent changeset of the stash */
+  readonly parentChangeset: ChangesetIdWithIndex;
+
+  /** ID sequences for the stash */
+  readonly idSequences: {
+    element: Id64String;
+    instance: Id64String;
+  };
+
+  /** Transaction properties for the stash */
+  readonly txns: TxnProps[];
+
+  /** Number of locks acquired by the stash */
+  readonly acquiredLocks: number;
+}
+
+/**
+ * Properties for creating a stash
+ * @internal
+ */
+export interface CreateStashProps {
+  /** Briefcase database instance */
+  readonly db: BriefcaseDb;
+  /** description of the stash */
+  readonly description: string;
+  /** discard all local changes and unless keepLocks flag is set, all locks will be released */
+  readonly discardLocalChanges?: true;
+  /** keep all locks after discarding local changes */
+  readonly retainLocks?: true;
+}
+
+/**
+ * Arguments for stash operations
+ * @internal
+ */
+export interface StashArgs {
+  readonly db: BriefcaseDb;
+  readonly stash: Id64String | StashProps;
+}
+
+/**
+ * Arguments for applying or restoring a stash.
+ *
+ * Extends {@link StashArgs} with an additional `method` property to specify the operation.
+ *
+ * @property method - Specifies the stash operation to perform.
+ *   - `"apply"`: Apply the stash without removing it.
+ *   - `"restore"`: Apply the stash and remove it from the stash list.
+ * @internal
+ */
+export interface StashApplyArgs extends StashArgs {
+  readonly method: "apply" | "restore";
```

Comment:
I'm trying to understand why you use a union of string literal here and a enum type for `LockOrigin`. Other than having 2 values vs 3 values, I don't really see a difference in the two use cases. If we go with the "string literal unions are preferred over enums" argument, then why have an enum for LockOrigin. If we go with the "enums are preferred over string literal unions" argument, then why does this use a string literal union?

Note: I honestly don't know if one of the two (string literal unions vs enums) is considered to be "better" in iTwin. However, even if we don't care, it seems a bit odd that these two (right next to each other in the code) answer the question with opposite answers.

Reply:
Forgot to mention: I do get that `LockOrigin` is a numeric enum, and not a string-based enum, but I don't think that really affects things (although I could be wrong).

Reply:
So i will remove this as well. Stash can be applied on top of local changes but its not part of this PR. This PR can only do restore usecase where briefcase is reverted to a previous state. So i am remove StashApplyArgs and will add it when we implement Stash apply on top of local change feature.